Remora.Discord.Extensions/Embeds/EmbedBuilder.cs

```diff
@@ -0,0 +1,384 @@
+﻿//
+//  EmbedBuilder.cs
+//
+//  Author:
+//       Jarl Gullberg <jarl.gullberg@gmail.com>
+//
+//  Copyright (c) 2017 Jarl Gullberg
+//
+//  This program is free software: you can redistribute it and/or modify
+//  it under the terms of the GNU Lesser General Public License as published by
+//  the Free Software Foundation, either version 3 of the License, or
+//  (at your option) any later version.
+//
+//  This program is distributed in the hope that it will be useful,
+//  but WITHOUT ANY WARRANTY; without even the implied warranty of
+//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+//  GNU Lesser General Public License for more details.
+//
+//  You should have received a copy of the GNU Lesser General Public License
+//  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+//
+
+using System;
+using System.Collections.Generic;
+using System.ComponentModel.DataAnnotations;
+using System.Drawing;
+using System.Linq;
+using Remora.Discord.API;
+using Remora.Discord.API.Abstractions.Objects;
+using Remora.Discord.API.Objects;
+using Remora.Discord.Extensions.Errors;
+using Remora.Rest.Core;
+using Remora.Results;
+
+namespace Remora.Discord.Extensions.Embeds
+{
+    /// <summary>
+    /// Provides utilities for building an embed.
+    /// </summary>
+    public class EmbedBuilder
+    {
+        /// <summary>
+        /// Gets or sets the title of the embed.
+        /// </summary>
+        [MaxLength(Constants.MaxTitleLength)]
```

Comment:
I'm unfamiliar with these annotations, so forgive me if I ask some obvious questions x)

Is there some analyzer or other mechanism that checks the validity of input or output data from members annotated with these attributes? If not, what's the use case for these over <remarks> in the doc header?

Reply:
These annotations live in the `System.ComponentModel.DataAnnotations` library. It looks like they were originally written for ASP.NET but were ported to be more generic with the advent of .NET Core. Either way, the analyzer in this case appears to be the .NET Runtime itself (through `System.Diagnostics.CodeAnalysis`). I'm not entirely familiar with the back-end system, but `MaxLengthAttribute` has an [`IsValid()`](https://source.dot.net/#System.ComponentModel.Annotations/System/ComponentModel/DataAnnotations/MaxLengthAttribute.cs,0e085bd95ad8e4c6,references) method which returns a boolean. I believe by default the system then throws an IOE or ArgumentOutOfRangeException.

Reply:
Added code to the `Ensure()` method to ensure these function.